spec/function.dd

```diff
@@ -2749,17 +2749,56 @@ $(H3 $(LNAME2 trusted-functions, Trusted Functions))
 
         $(P Trusted functions are marked with the $(CODE @trusted) attribute.)
 
-        $(P Trusted functions are guaranteed by the programmer to not exhibit
-        any undefined behavior if called by a safe function.
-        Generally, trusted functions should be kept small so that they are
-        easier to manually verify.
+        $(P Trusted functions are guaranteed to not exhibit any undefined
+        behavior if called by a safe function. Furthermore, calls to trusted
+        functions cannot lead to undefined behavior in @safe code that is
+        executed afterwards. It is the responsibility of the programmer to
+        ensure that these guarantees are upheld.
         )
 
+        $(P Example:)
+
+        ---
+        immutable(int)* f(int* p) @trusted
+        {
+            version (none) p[2] = 13;
+            // Invalid. p[2] is out of bounds. This line would exhibit undefined
+            // behavior.
+
+            version (none) p[1] = 13;
+            // Invalid. In this program, p[1] happens to be in-bounds, so the
+            // line would not exhibit undefined behavior, but a trusted function
+            // is not allowed to rely on this.
```

Comment:
I'm not sure I understand the utility of this example. What if the user passes the length of `p` as a separate integral parameter? That would be good, right?

So for this example it would be "valid if and only if the actual extent of `p` is two or more elements".

If I understand correctly, do you want to limit `@trusted` functions taking pointers to accessing at most ONE element pointed to by those pointers? That would be a large restriction.

Reply:
> I'm not sure I understand the utility of this example. What if the user passes the length of `p` as a separate integral parameter? That would be good, right?

No. That's not allowed either. Taking a D array is where it's at. Then it's guaranteed that the length is valid.

> So for this example it would be "valid if and only if the actual extent of `p` is two or more elements".

Exactly not that. In your version, it's possible to call an `@trusted` function in an unsafe way  from `@safe` code. In my opinion, that's what we want to prevent.

> If I understand correctly, do you want to limit `@trusted` functions taking pointers to accessing at most ONE element pointed to by those pointers?

Yes, and I don't think it's a new limitation. It's the reason why many C functions, e.g. `memcpy`, aren't marked as `@trusted` in core.stdc. But we can write an `@trusted` wrapper that takes D arrays and calls `memcpy` with a known-good length.

Reply:
The working rule here is that an @trusted function must present a safe interface. Therefore, any pointer parameter that will perform arithmetic on the parameter and then dereference it cannot be @trusted, it must be @system. Your example and rationale with `memcpy` is correct.